### PR TITLE
Add context support, ErrUnauthorized and Doer interface for http.Client

### PR DESCRIPTION
**Pull request by anpryl**
Extend API for context usage.
Make Errors an error instance.

### CODE REVIEW

File: asana/asana.go
Comment (shurcooL) [inline asana/asana.go]: Do these 3 `fmt.Println`s belong here? Did you forget to remove them while testing?

File: asana/asana.go
Comment (dmitshur) [inline asana/asana.go]: Did you forget to remove this change after testing?

**Review by dmitshur — COMMENTED** [review]
Thank you for contributing.

Aside from the `fmt.Println(resp.StatusCode, string(b))` change, which looks unintended, all other code changes look good.

However, I see two main issues/potential improvements for this PR.
1. All code changes are in a single commit. This is harder to review, and not a good practice. Would you mind splitting each logical change like "add context support, "add Doer interface", "add http post form", etc., into a separate commit?
2. No documentation. I know a lot of the code here isn't perfect, but let's try to make things better. Exported symbols need to be documented. This is a good practice and consistent with Go style. Try running `golint` on your changes.

Aside from those 2 things, these changes look great and make sense to me. I'd be happy to accept them. I'm guessing @tambet would like them too.

Are you willing to address the two suggestions I've made above?

File: asana/asana.go
Comment (anpryl) [reply asana/asana.go]: I'm so sorry for this debugging printing
I will add documentation and split commits at this weekend.
Thank you for your feedback!

**Comment by anpryl** [comment]
Tried to split commits as much as possible, add documentation, also add tests this time.
If I could make it better, please, don't hesitate to comment it.
Thank you!

File: asana/asana.go
Comment (dmitshur) [inline asana/asana.go:47]: Here, and elsewhere, the comments don't have a space after the two slashes. There should be a single space there.

See [here](https://dmitri.shuralyov.com/idiomatic-go#comments-for-humans-always-have-a-single-space-after-the-slashes) for rationale.

File: asana/asana.go
Comment (dmitshur) [inline asana/asana.go:43]: I have a guess, but could you please elaborate on why you decided to make the `Doer` interface instead of sticking with an `*http.Client`?

File: asana/asana.go
Comment (dmitshur) [inline asana/asana.go:146]: Are there any restrictions on `Errors`? Can it contain 0 elements? 1 element?

File: asana/asana.go
Comment (dmitshur) [inline asana/asana.go:315]: According to this, at least 1 element. I think it's worth documenting that for the `Errors` type, so users will know to always expect either at least one error there.

File: asana/asana_test.go
Comment (dmitshur) [inline asana/asana_test.go]: I know it's a test, but still, this isn't saving much typing. Instead of this variable here, just use `context.Background()` directly at the call sites. E.g.:

``` Go
workspaces, err := client.ListWorkspaces(context.Background())
...
users, err := client.ListUsers(context.Background(), nil)
...
```

File: asana/asana.go
Comment (dmitshur) [inline asana/asana.go:262]: Now `request` has 2 ways of specifying the data. Either by setting `data`, or by setting `form`. I think it might be helpful to document that at most one can be used at the same time (or none), and that `data` takes higher precedence.

What do you think?

**Review by dmitshur — CHANGES_REQUESTED** [review]
This looks great. I've left minor comments and questions in my review. Thanks!

File: asana/asana.go
Comment (anpryl) [reply asana/asana.go:43]: It allows me to use single instance of my configured http.Client and set different Auth headers on each request.

I have configured http.Client and it is property of my own Client struct.

``` go
type Client struct {
    httpClient *http.Client
}
```

My service receiving auth token as an argument to request and with Doer interface it is easy to add Auth header with token to request like this:

``` go
func (c *Client) withToken(token string) *asana.Client {
    f := func(req *http.Request) (*http.Response, error) {
        req.Header.Set("Authorization", "Bearer "+token)
        return c.httpClient.Do(req)
    }
    return asana.NewClient(asana.DoerFunc(f))
}

// Usage example
func (c *Client) Issue(ctx context.Context, token string, issueID int64) (*Issue, error) {
    t, err := c.withToken(token).GetTask(ctx, issueID, issueFieldsFilter)
    if err != nil {
        return nil, err
    }
    issue := toIssue(t)
    return &issue, nil
}
```

File: asana/asana.go
Comment (anpryl) [reply asana/asana.go:47]: Fix this, thank you for explanations.

File: asana/asana.go
Comment (anpryl) [reply asana/asana.go:262]: Improve request documentation

File: asana/asana.go
Comment (anpryl) [reply asana/asana.go:315]: You are correct, add godoc for Errors type

File: asana/asana_test.go
Comment (anpryl) [reply asana/asana_test.go]: I prefer to write 3 letters instead of context.Background() in tests, but it's not a big deal. I remove this global variable and put context.Background().

File: asana/asana.go
Comment (dmitshur) [reply asana/asana.go:43]: Thanks for the explanation.

This sounds reasonable, but I also slightly worry this might not be the most optimal solution for the problem you're trying to solve. Having `Doer` be an interface lets you pass in a middleware-augmented `*http.Client` of sorts, but you still have to create a new `asana.NewClient` for each one...

It's also interesting that go-github, a very popular library, manages to get by without `Doer` interface. /cc @willnorris, do you know how one would do that with go-github?

I'm not opposed to this going in @anpryl, I just wanted to say that it seems a little more questionable to me, and perhaps we can find a better way to effectively be able to augment certain `*http.Request`s. But maybe not, and this is fine.

**Review by dmitshur — APPROVED** [review]
Thanks for addressing all my feedback @anpryl.

This LGTM.

@tambet, any comments on this? If this looks good to you, or if you don't have any objections, I'll merge this in a few days.

File: asana/asana.go
Comment (willnorris) [reply asana/asana.go:43]: Is this client fully taking on responsibility for authentication, or only for certain calls, like in your c.Issue method?  go-github's Client knows nothing about authentication at all.  It is always punted to the underlying http.Client, which typically comes from golang.org/x/oauth2, or the basic auth client that go-github does provide (but is otherwise completely separated from the Client struct).

I think the best analogy to this in go-github would be the custom accept headers we set on various API calls.  That happens by first building the http.Request, then adding whatever custom headers we want before passing it off to client.Do.  As long as the underlying http.Client leaves those headers intact, then you're fine.  If your underlying http.Client is clobbering your auth header because it thinks it should be responsible for auth, then you do have a problem.  

This is exactly why go-github's Client struct has its own Do and NewRequest methods, so that you can modify the request before calling Do.  Not separating those, as go-asana does with its single Request method, results in these weird workarounds.

Comment (dmitshur) [reply asana/asana.go:43]: Thanks for the explanation @willnorris, that makes sense.

I agree that the `Do` and `NewRequest` separation makes more sense and is cleaner/safer, but I'd want to think more and understand it better (outside of this PR). I think we should consider changing this library in that direction, rather than the `Doer`, which I think is a lot harder for people to use (at all, and also correctly).

What are your thoughts on this @anpryl?

Comment (willnorris) [reply asana/asana.go:43]: If you're concerned about the API surface, you don't have to export those methods either.  I made the conscious decision to export them in go-github, but you don't necessarily need to, since you just need them for internal use.  If you went that route, you'd probably want to refactor the existing `request` method to use the new `newRequest` and `do` methods, but it wouldn't change the API at all.  Just a thought.

Comment (dmitshur) [reply asana/asana.go:43]: @willnorris Hmm, but if it's not exported API, how would @anpryl be able to achieve his original goal?

> It allows me to use single instance of my configured http.Client and set different Auth headers on each request.

Comment (anpryl) [reply asana/asana.go:43]: @shurcooL you are correct, I could not be able to achieve my goals with this approach.
I take a look at go-github and didn't find way to achieve my goal too.

I have another proposition about `NewClient`. Move parsed `*url.URL` from line 134 to package level and parse it in `init()` function. It will decrease cost of `asana.Client` initialization with NewClient function.

Comment (willnorris) [reply asana/asana.go:43]: Oh, I must have misunderstood that part.  I thought the per-request auth header was still being set inside the asana package, which would have access to the non-exported methods.  If those auth headers are being set by the caller of this package then yes you're right, they'd need to be exported.

Comment (dmitshur) [reply asana/asana.go:43]: > I have another proposition about `NewClient`. Move parsed `*url.URL` from line 134 to package level and parse it in `init()` function. It will decrease cost of `asana.Client` initialization with NewClient function.

@anpryl, that sounds very reasonable. Would that mean we won't need to include the `Doer` interface in this PR?

I'm still open to discussing it further, but it seems better to do that outside of this PR, since it already does a lot and has taken a while. I don't want to make you wait even longer. :)

Comment (anpryl) [reply asana/asana.go:43]: We still need `Doer` to reuse `http.Client` with different tokens.
But this change will decrease cost of `NewClient` function call.
We could also include `withToken` function into package and make it public.

Comment (dmitshur) [reply asana/asana.go:43]: Well, okay. It's been a while and I have no better ideas than `Doer` interface, so let's go with that for now. If really good ideas come in the future, we can probably still change things.

**Review by dmitshur — APPROVED** [review]
LGTM. See https://github.com/tambet/go-asana/pull/4#discussion_r83365539.

I'll let this sit for a day and merge tomorrow if no further comments. /cc @tambet

Thanks for your work @anpryl!

File: asana/asana.go
Comment (anpryl) [reply asana/asana.go:43]: Can I move `url.URL` to global variable and initialize it in `init()` funciton?

Comment (dmitshur) [reply asana/asana.go:43]: It's better to split up orthogonal changes across multiple PRs and keep each one smaller.

Mind making it a separate followup PR? This one is already huge. :)

Comment (anpryl) [reply asana/asana.go:43]: Well, you are right, I will prepare new PR after merge.

**Comment by dmitshur** [comment]
Merged. This is a breaking API change, but it includes some necessary API improvements such as adding support for passing a context. Updating should be easy, and those who don't want to will have this library vendored anyway.

Thanks for your awesome contribution @anpryl!

**Pull request merged**
